### PR TITLE
Add support for Array subqueries

### PR DESCRIPTION
**Pull request by jkosh44**
Postgres allows array construction through a query, called Array
subqueries. The syntax is `ARRAY(<query>)`. The result is an array
where each element is a row from the query provided. This commit
implements this feature. Currently it has the limitation that the
result of the query itself can't be an array. The reason is that array
subqueries use the array_agg function, which doesn't support nested
arrays.

This commit was modeled heavily off of #8571, which implements
the same feature for Lists.

### Motivation

This PR adds a feature that has not yet been specified, see above. 

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by jkosh44** [comment]
Functionality for this is done, I'm just trying to see where I can eliminate duplication between this and List subqueries. I gave up for now on supporting nested Arrays and instead opened this issue: https://github.com/MaterializeInc/materialize/issues/10701

**Review by sploiselle — COMMENTED** [review]
Talked to Joe about this and he's well underway to generalizing this implementation w/ the LIST subquery code, so going to hold off on really digging into this until he has more of that work done.

**Comment by jkosh44** [comment]
@sploiselle Should be ready for review now, I've de-duplicated the code between array subquery and list subquery planning. Not super happy about the names I came up with so feel free to suggest better ones.

File: src/sql-parser/tests/testdata/select [hunk 1]
```diff
@@ -1199,6 +1199,51 @@ SELECT LIST(WITH usps AS (SELECT 42) SELECT LIST[customer.id, LIST[customer.firs
...
+----
+SELECT ARRAY(SELECT customer.id FROM customer JOIN user ON customer.id = user.id LIMIT 12)
+=>
+Select(SelectStatement { query: Query { ctes: [], body: Select(Select { distinct: None, projection: [Expr { expr: ArraySubquery(Query { ctes: [], body: Select(Select { distinct: None, projection: [Expr { expr: Identifier([Ident("customer"), Ident("id")]), alias: None }], from: [TableWithJoins { relation: Table { name: Name(UnresolvedObjectName([Ident("customer")])), alias: None }, joins: [Join { relation: Table { name: Name(UnresolvedObjectName([Ident("user")])), alias: None }, join_operator: Inner(On(Op { op: Op { namespace: [], op: "=" }, expr1: Identifier([Ident("customer"), Ident("id")]), expr2: Some(Identifier([Ident("user"), Ident("id")])) })) }] }], selection: None, group_by: [], having: None, options: [] }), order_by: [], limit: Some(Limit { with_ties: false, quantity: Value(Number("12")) }), offset: None }), alias: None }], from: [], selection: None, group_by: [], having: None, options: [] }), order_by: [], limit: None, offset: None }, as_of: None })
+
+parse-statement
+SELECT ARRAY(
+  WITH usps AS (SELECT 42)
```
Comment (sploiselle) [inline src/sql-parser/tests/testdata/select:1233]: We should also include that the CTE can be referenced; is this from a parallel `list` test?

File: test/sqllogictest/array_subquery.slt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+# Verify nested arrays
+query error ARRAY expression cannot be cast to uniform type: text\[\] vs integer\[\]
+SELECT ARRAY[ARRAY[customer.first_name, customer.last_name], ARRAY[customer.zip], ARRAY[customer.id]]::text FROM customer JOIN users ON customer.id = users.id ORDER BY users.other_field DESC LIMIT 2
+
+query error text\[\]\[\] not yet supported
+SELECT ARRAY(SELECT ARRAY[customer.first_name] FROM customer)
+
+# Check CTE syntax can be included in a query
+query T
+SELECT ARRAY(WITH usps AS (SELECT 42) SELECT customer.first_name FROM customer)
```
Comment (sploiselle) [inline test/sqllogictest/array_subquery.slt:146]: ditto comment about reference; maybe that doesn't matter, as much, or is intentionally testing this behavior?

**Review by sploiselle — APPROVED** [review]
LGTM modulo comments about tests, which I'd like to see fixed

File: test/sqllogictest/array_subquery.slt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+query T
+SELECT ARRAY(SELECT other_field FROM users ORDER BY id ASC)::text
+----
+{10,5,8}
+
+query error Expected subselect to return 1 column, got 2 columns
+SELECT ARRAY(SELECT first_name, last_name FROM customer)::text
+
+# Verify nested arrays
+query error ARRAY expression cannot be cast to uniform type: text\[\] vs integer\[\]
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/sqllogictest/array_subquery.slt]: Agreed, I moved this to `arrays.slt`

File: test/sqllogictest/array_subquery.slt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+SELECT ARRAY(SELECT x FROM xs WHERE x > 1)::text
+----
+{2}
+
+query T
+SELECT ARRAY[ARRAY(SELECT x FROM xs)]::text
+----
+{{1,1,2}}
+
+statement ok
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/sqllogictest/array_subquery.slt]: Changed to query

File: src/sql-parser/tests/testdata/select (hunk 1 above)
Comment (jkosh44) [reply src/sql-parser/tests/testdata/select:1243]: I added a second test that references the CTE

File: src/sql/src/plan/query.rs [hunk 5]
```diff
@@ -3330,12 +3394,14 @@ fn plan_list_subquery(
         .unwrap()
         .scalar_type();
 
+    if is_unsupported_type(&elem_type) {
+        bail_unsupported!(format!("{}[]", ecx.humanize_scalar_type(&elem_type)));
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/sql/src/plan/query.rs]: Just added details to the error

**Pull request merged**